### PR TITLE
[TTS] Add STFT and SI-SDR loss to audio codec recipe

### PR DESCRIPTION
**Pull request by rlangman**
# What does this PR do ?

Make audio codec losses more configurable, and add STFT and SI-SDR loss definitions.

**Collection**: [TTS]

# Changelog 
- Rename encodec.yaml to encodec_24000.yaml to include sampling rate, and modify parameters to better match the original paper.
- Add audio_codec_24000.yaml where we can start creating config for optimized NeMo audio codec recipe. 
- Separate out aggregate mel loss into separate L1 and L2 losses.
- Add parameter to control mel dimension for different resolutions.
- Implement multi-resolution log STFT loss.
- Implement SI-SDR loss, being a slightly modified version of the torchmetrics implementation but with masking support.
- Make all losses optional during train_step, only computing and optimizing on them if their loss_scale is non-zero. Except L1 mel loss which I expect will always be used, so I can use it as the logged "t_loss" value.
- Compute all losses during validation step for monitoring and validation.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

# Additional Information
The new configuration places no weight on time domain loss, scales the mel dimension proportional to the resolution (like DAC), and adds STFT loss (like univnet) to further improve accuracy. 

These are what the validation losses look like when training with different losses: mel fixed dimension, mel scaled dimension, STFT, and (mel scaled & STFT).

loss_mel

loss_si_sdr

loss_time_domain

loss_stft

### CODE REVIEW

File: examples/tts/conf/audio_codec/audio_codec_24000.yaml [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+val_ds_meta: ???
+
+log_ds_meta: ???
+log_dir: ???
+
+# Modify these values based on your sample rate
+sample_rate: 24000
+train_n_samples: 24000
+down_sample_rates: [2, 4, 5, 8]
+up_sample_rates: [8, 5, 4, 2]
```
[the remark this answers is not in the document]
Comment (rlangman) [reply examples/tts/conf/audio_codec/audio_codec_24000.yaml:26]: It usually will be, but does not have to be. We do plan to run some experiments with asymmetric encoder/decoder, or encode spectrogram instead of raw audio (for which there will be no `down_sample_rates`). Whether those configurations will be useful has yet to be seen.

**Review by redoctopus — COMMENTED** [review]
LGTM, but I want to wait for at least one other person to take a look.
